x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportSearchableSnapshotsStatsAction.java

```diff
@@ -0,0 +1,144 @@
+    private final IndicesService indicesService;
+
+    @Inject
+    public TransportSearchableSnapshotsStatsAction(ClusterService clusterService, TransportService transportService,
+                                                   IndicesService indicesService, ActionFilters actionFilters,
+                                                   IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(SearchableSnapshotsStatsAction.NAME, clusterService, transportService, actionFilters, indexNameExpressionResolver,
+            SearchableSnapshotsStatsRequest::new, ThreadPool.Names.MANAGEMENT);
+        this.indicesService = indicesService;
+    }
+
+    @Override
+    protected ClusterBlockException checkGlobalBlock(ClusterState state, SearchableSnapshotsStatsRequest request) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected ClusterBlockException checkRequestBlock(ClusterState state, SearchableSnapshotsStatsRequest request, String[] indices) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, indices);
+    }
+
+    @Override
+    protected SearchableSnapshotStats readShardResult(StreamInput in) throws IOException {
+        return new SearchableSnapshotStats(in);
+    }
+
+    @Override
+    protected SearchableSnapshotsStatsResponse newResponse(SearchableSnapshotsStatsRequest request,
+                                                           int totalShards, int successfulShards, int failedShards,
+                                                           List<SearchableSnapshotStats> searchableSnapshotStats,
+                                                           List<DefaultShardOperationFailedException> shardFailures,
+                                                           ClusterState clusterState) {
+        return new SearchableSnapshotsStatsResponse(searchableSnapshotStats, totalShards, successfulShards, failedShards, shardFailures);
+    }
+
+    @Override
+    protected SearchableSnapshotsStatsRequest readRequestFrom(StreamInput in) throws IOException {
+        return new SearchableSnapshotsStatsRequest(in);
+    }
+
+    @Override
+    protected ShardsIterator shards(ClusterState state, SearchableSnapshotsStatsRequest request, String[] concreteIndices) {
+        final List<String> searchableSnapshotIndices = new ArrayList<>();
+        for (String concreteIndex : concreteIndices) {
+            IndexMetaData indexMetaData = state.metaData().index(concreteIndex);
+            if (indexMetaData != null) {
+                Settings indexSettings = indexMetaData.getSettings();
+                if (INDEX_STORE_TYPE_SETTING.get(indexSettings).equals(SNAPSHOT_DIRECTORY_FACTORY_KEY)) {
+                    searchableSnapshotIndices.add(concreteIndex);
+                }
+            }
+        }
+        return state.routingTable().allShards(searchableSnapshotIndices.toArray(new String[0]));
```

Comment:
Ideally, we would throw an exception if the user specified an index (or pattern) which didn't match any searchable snapshots, but this would naturally fall to the `IndexNameExpressionResolver` which doesn't seem to have a suitable extension point for this kind of logic.

However, what do you think about handling at least the case of a user specifying a single index with a typo, matching nothing, with an INFE? I.e. if `request.indices()` is not empty or `["*"]` or `["_all"]` but `searchableSnapshotIndices` is empty, then that's a bad request.

Reply:
That sounds like a good idea. I implemented it a bit differently though in c9468d104a8f341b94ba4bbba8373f68f30b508b, by checking if one or more concrete indices were resolved (taking security into account) but none of them had a `searchable` store type (`searchableSnapshotIndices` is empty) which throws a resource not found exception. I've updated the REST API test for this.

Reply:
Hmm, this means that `GET _searchable_snapshots/stats` returns `200 OK` in an empty cluster but `404 Not Found` if there are any indices at all but none of them are searchable snapshots. I think that'll be surprising. Maybe it'd be best always to throw a RNFE if `searchableSnapshotIndices` is empty.

Reply:
I'm not fully sure how it should behave, also because Security also has its specific behavior so I went with your suggestion in d1049a35844f78f1faa595e01745efe472b5acbf which has the advantage to be consistent.

Reply:
Looks good. Test failures are not obviously related, maybe we need to merge something?

Reply:
Yes, need to merge master but I'm waiting for bwc to be reenabled before.